### PR TITLE
fix(otel): Exporter creating Monitored Resource with task_id for Cloud Run

### LINKED ISSUE #14925: OpenTelemetry on Cloud Run does not include `task_id` in Monitored Resource
**What component of `google-cloud-cpp` is this related to?**: `google/cloud/opentelemetry`

**Describe the bug**: When inside a Cloud Run environment, the [`MonitoredResource`](https://cloud.google.com/monitoring/api/ref_v3/rpc/google.monitoring.v3#google.monitoring.v3.TimeSeries) in a [`CreateTimeSeriesRequest`](https://cloud.google.com/monitoring/api/ref_v3/rpc/google.monitoring.v3#google.monitoring.v3.CreateTimeSeriesRequest) to the Cloud Monitoring API does not include the necessary fields for the [`generic_task`](https://cloud.google.com/monitoring/api/resources#tag_generic_task) resource type, and is rejected.

**To Reproduce**

* Deploy a binary on Cloud Run that uses `MakeResourceDetector()` to create an OTel Resource object, and uses `MakeMonitoringExporter()` to export to Cloud Monitoring
* Observe errors in Cloud Monitoring audit logs: 
> One or more TimeSeries could not be written: timeSeries[1]: The set of resource labels is incomplete. Missing labels: (namespace task_id).; timeSeries[0]: The set of resource labels is incomplete. Missing labels: (namespace task_id).; timeSeries[2]: The set of resource labels is incomplete. Missing labels: (namespace task_id).

**Expected behavior**: Should follow the [well-tested Golang implementation](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/8da0f42dab085c916987891419461d583a2aa96e/internal/resourcemapping/resourcemapping.go#L153) where the `faas.instance` OTel Resource Attribute is mapped to `MonitoredResource` `task_id`. As the `service.namespace` OTel Resource Attribute is not set by the Resource Detector from within Cloud Run, it should be mapped as an empty string (as in [golang implementation](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/8da0f42dab085c916987891419461d583a2aa96e/internal/resourcemapping/resourcemapping.go#L250)), rather than being left absent.

**What version of `google-cloud-cpp` are you using?**: `2.3.0`

**Example OTel Resource Attributes from within Cloud Run**:

```
service.name: unknown_service
telemetry.sdk.name: opentelemetry
faas.name: my-cloud-run-service-name
telemetry.sdk.version: 1.16.1
cloud.platform: gcp_cloud_run
telemetry.sdk.language: cpp
faas.instance: 0066d92480fc484796ce902f1b607e033bc3750442848d702f657a75e12c4b4dd89466b825feca5caf88fda7618a27794c7f4724cabd17de134aadde1d2e8abbf124fc5d2f
faas.version: my-cloud-run-service-name-00076-2lq
cloud.account.id: my-project-name
cloud.provider: gcp
```
 
I have a draft fix in pull request #14923

### PR DESCRIPTION
**Pull request by DouglasHeriot**
Fixes #14925

When inside a Cloud Run environment, the [`MonitoredResource`](https://cloud.google.com/monitoring/api/ref_v3/rpc/google.monitoring.v3#google.monitoring.v3.TimeSeries) in a [`CreateTimeSeriesRequest`](https://cloud.google.com/monitoring/api/ref_v3/rpc/google.monitoring.v3#google.monitoring.v3.CreateTimeSeriesRequest) to the Cloud Monitoring API does not include the necessary fields for the [`generic_task`](https://cloud.google.com/monitoring/api/resources#tag_generic_task) resource type, and is rejected.

Also updating the GenericNode mapping to match Golang.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14923)

### CODE REVIEW

File: google/cloud/opentelemetry/internal/monitored_resource.cc [hunk 1]
```diff
@@ -74,6 +74,8 @@ class MonitoredResourceProvider {
         mr.labels[kv.first] = AsString(attributes.at(*found));
       } else if (kv.second.fallback) {
         mr.labels[kv.first] = *kv.second.fallback;
+      } else {
+        mr.labels[kv.first] = "";
```
Comment (dbolduc) [inline google/cloud/opentelemetry/internal/monitored_resource.cc:78]: So it seems like we always fallback. Sometimes to a value like `"global"`, but at least to `""`.

optional nit: It seems slightly cleaner to make `fallback` a `std::string`:

https://github.com/googleapis/google-cloud-cpp/blob/644efe9cf17e4a29397aa326e9fce98e4792e3e8/google/cloud/opentelemetry/internal/monitored_resource.cc#L55

and then have this code just be:

```cc
} else {
  mr.labels[kv.first] = kv.second.fallback;
}
```

(If this doesn't work for whatever reason, just let me know, and I can look into cleaning it up in a future PR)

File: google/cloud/opentelemetry/internal/monitored_resource.cc [hunk 2]
```diff
@@ -174,7 +176,7 @@ MonitoredResourceProvider GenericNode() {
           {"location",
            {{sc::kCloudAvailabilityZone, sc::kCloudRegion}, "global"}},
           {"namespace", {{sc::kServiceNamespace}}},
-          {"node_id", {{sc::kHostId}}},
+          {"node_id", {{sc::kHostId, sc::kHostName}}},
```
Comment (dbolduc) [inline google/cloud/opentelemetry/internal/monitored_resource.cc:179]: Good catch, thanks!

Can we have a unit test for this?

https://github.com/googleapis/google-cloud-cpp/blob/90d72c0e6d81d9e54cdc1faf2f5044832d59997c/google/cloud/opentelemetry/internal/monitored_resource_test.cc#L337

I would probably change `TestCase` to something like `LocationTestCase`, and then introduce a `NodeIdTestCase`, and loop over:

```cc
for (auto l : location_tests) {
  for (auto n : node_tests) {
    // verify l.expected_location and n.expected_node_id
  }
}
```

File: google/cloud/opentelemetry/internal/monitored_resource.cc [hunk 3]
```diff
@@ -162,8 +164,8 @@ MonitoredResourceProvider GenericTask() {
           {"location",
            {{sc::kCloudAvailabilityZone, sc::kCloudRegion}, "global"}},
           {"namespace", {{sc::kServiceNamespace}}},
-          {"job", {{sc::kServiceName}}},
-          {"task_id", {{sc::kServiceInstanceId}}},
+          {"job", {{sc::kServiceName, sc::kFaasName}}},
+          {"task_id", {{sc::kServiceInstanceId, sc::kFaasInstance}}},
```
Comment (dbolduc) [inline google/cloud/opentelemetry/internal/monitored_resource.cc:168]: Again, good catch and thanks.

Can we have unit tests for these too? Or at least fix the broken tests.

Also, consider leaving a comment in a test case where we fallback to the empty string. Just something that says "Verify that we fallback to the empty string if no matches are found".

**Review by dbolduc — COMMENTED** [review]
Thanks for the PR!

The code looks good, although I offered an optional refactor. Unit tests for the new code paths would be nice. Either way, we need to fix the tests that are now broken.

You can run the unit test with:

```
bazelisk test --test_output=all //google/cloud/opentelemetry:internal_monitored_resource_test
```

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 1 above)
Comment (DouglasHeriot) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:78]: That does seem cleaner, I’ll change the fallback to default to empty string.

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 3 above)
Comment (DouglasHeriot) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:168]: Yep, I’ll add some tests for these cases.

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 2 above)
Comment (DouglasHeriot) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:177]: Done

File: google/cloud/opentelemetry/internal/monitored_resource.cc [hunk 4]
```diff
@@ -52,7 +52,7 @@ struct AsStringVisitor {
 
 struct OTelKeyMatch {
   std::vector<std::string> otel_keys;
-  absl::optional<std::string> fallback = absl::nullopt;
+  std::string fallback;
```
Comment (dbolduc) [inline google/cloud/opentelemetry/internal/monitored_resource.cc:55]: might need to explicitly say `std::string fallback = {}` for our initializer lists to work

**Review by dbolduc — COMMENTED** [review]
To lint the code:

```
ci/cloudbuild/build.sh -t checkers-pr
```

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 1 above)
Comment (dbolduc) [reply google/cloud/opentelemetry/internal/monitored_resource.cc]: Thanks!

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 4 above)
Comment (DouglasHeriot) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:55]: I think this works already – `std::string` default constructs to empty string and the compiler doesn’t complain?

**Comment by DouglasHeriot** [comment]
Sorry about the formatting, I think it’s fixed now.

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 4 above)
Comment (dbolduc) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:55]: So I am not smart enough to tell you why the compiler is complaining. (and only in cmake-based builds :shrug:). But it is complaining:

image

e.g. this build: https://github.com/googleapis/google-cloud-cpp/pull/14923/checks?check_run_id=35680003918

Comment (dbolduc) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:55]: I think the simplest build to test on is:

```
ci/cloudbuild/build.sh -t demo-fedora-pr
```

Comment (dbolduc) [reply google/cloud/opentelemetry/internal/monitored_resource.cc:55]: It looks like we had to do the same thing (explicitly initialize the field to its default value) when the type was `absl::optional<std::string>`. :shrug:

Comment (DouglasHeriot) [reply google/cloud/opentelemetry/internal/monitored_resource.cc]: Ah you’re right. Fixed now. Thanks for the build command.

**Review by dbolduc — APPROVED** [review]
Thanks!

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 4 above)
Comment (DouglasHeriot) [reply google/cloud/opentelemetry/internal/monitored_resource.cc]: Haha, `clang-tidy` doesn’t like this:

```
error: initializer for member 'fallback' is redundant [readability-redundant-member-init,-warnings-as-errors]
Step #2:    55 |   std::string fallback = {};
```

I guess I’ll have to find the syntax to make it ignore this warning? Probably a better solution than ignoring the above `-Wmissing-field-initializers` warnings

File: google/cloud/opentelemetry/internal/monitored_resource.cc [hunk 5]
```diff
@@ -52,7 +52,7 @@ struct AsStringVisitor {
 
 struct OTelKeyMatch {
   std::vector<std::string> otel_keys;
-  absl::optional<std::string> fallback = absl::nullopt;
+  std::string fallback = {};
```
Comment (dbolduc) [inline google/cloud/opentelemetry/internal/monitored_resource.cc:55]: Ugh, can you add a `NOLINT`, and I'll clean this up on my own time in a follow up PR?

```suggestion
  // NOLINTNEXTLINE(readability-redundant-member-init)
  std::string fallback = {};
```

I have no idea why clang-tidy did not complain for the previous code. Whatever.

File: google/cloud/opentelemetry/internal/monitored_resource.cc (hunk 4 above)
Comment (dbolduc) [reply google/cloud/opentelemetry/internal/monitored_resource.cc]: Uuuuuuuggghh...... I just added a commit to silence the compiler warning. I will deal with it later.[^1]

[^1]: if I feel like it.
